GPU profiling challenge using nsight systems

secondspass opened the pull request:
This challenge was based on some of the homework code from the OLCF CUDA training series. Let me know if you have any feedback. 

Also, let me know if this is better classified as a Medium or an Easy challenge (it's set to Hard right now because an earlier draft of this challenge was more involved).

secondspass committed:
update gitignore

secondspass committed:
started GPU profiling challenge, created Makefile and code

secondspass committed:
created submit scripts

secondspass committed:
GPU_Profiling README mostly finished

secondspass committed:
Fixed based on Tom's feedback

- Updated everything (readme, makefiles, batch scripts) to use Nsight systems
- De-emphasized the use of the GUI tools
- Added references

tom-papatheodore commented:
Overall, this looks great, but I have some suggestions...

* This challenge needs to be run on GPFS because the profiles are written from compute nodes. I believe all the other challenges assume they're running out of a home directory, so I would suggest (in the batch script) copying the executable to the user's GPFS scratch directory, cd-ing to the scratch directory to run, then copying the resulting files back to the working directory. Perhaps something like this:

```
cp ./matrix_sums_unoptimized /gpfs/wolf/gen146/scratch/${USER}
cd /gpfs/wolf/gen146/scratch/${USER}

jsrun -n1 -c1 -g1 -a1 nsys profile --stats=true --force-overwrite=true -o profile_data_unoptimized ./matrix_sums_unoptimized

cp ./profile_data_unoptimized* $LS_SUBCWD
```

* You include a PDF, but that file will not be able to open by clicking the link (it would have to be scp-d). I would suggest just pointing to where you found it online.

* When you describe the details of the kernels, I would state that it's not necessarily important to fully understand the code. It's sufficient to understand the general idea of why you get a performance increase - since the main point is about the profiler.

* I see there are URLs in the Useful Resources section but they do not link.

suzannepk commented:
Here is my feedback on GPU profiling: 
Overall well done! 

The main thing I struggled with was whether I should focus my energies on learning the parallel sweep reduction or learning how to use the profiler. Hard to separate because you must have an optimization technique that you believe works in mind to develop the intuition to trust that what the profile is telling you is useful. I think you have covered this with your initial explanation  of column_sums and coalesced memory. So for this tutorial, emphasizing the profiler is the way to go. 

Emphasize the profilers could mean things like this: 

**Tweaking the language:** 
There is a way to make the row_sums as fast as the column_sums, by using parallel sweep reduction. Look at the code in matrix_sums_optimized.cu and compare the row_sums kernel to the one in the matrix_sums_unoptimized.cu code.
Becomes
If you want to see an example of this optimization technique, compare the code in matrix_sums_optimized.cu and compare the row_sums kernel to the one in the matrix_sums_unoptimized.cu code. It may be helpful to open two terminal windows to compare results. 

**Adding more discussion of the profiles:** 
You could talk about other things you see in the profile- Discuss: Why does it take so much longer to copy the data from the host to the device than from the device to the host?  
I did have to edit the scripts to write to and from GPS as Tom suggested. 

Perhaps a good future addition would be breaking this into two challenges: 1 guiding them to code the simplest of parallel sweep reductions and 2. then doing the profiling on the example code you use here.

secondspass committed:
Updated based on Tom and Suzanne's feedback

- Deemphasized the parallel reduction part to focus more on profiling
- Added and corrected links
- Reduced level of detailed explanation, pointing to external sources instead
- Updated the batch scripts to copy to and from GPFS
- Removed the bonus points section because it doesn't really fit well with the theme

tom-papatheodore reviewed (approved):
LGTM!
